src/server/reset_to_version.rs

```diff
@@ -0,0 +1,379 @@
+        write_iter.seek(SeekKey::Start).unwrap();
+        lock_iter.seek(SeekKey::Start).unwrap();
+        Self {
+            write_iter,
+            lock_iter,
+            ts,
+            state,
+        }
+    }
+
+    fn next_write(&mut self) -> Result<Option<(Vec<u8>, Write)>> {
+        if self.write_iter.valid().unwrap() {
+            let mut state = self
+                .state
+                .lock()
+                .expect("failed to lock ResetToVersionWorker::state");
+            debug_assert!(matches!(
+                *state,
+                ResetToVersionState::RemoveWrite { scanned: _ }
+            ));
+            *state.scanned() += 1;
+            drop(state);
+            let write = box_try!(WriteRef::parse(self.write_iter.value())).to_owned();
+            let key = self.write_iter.key().to_vec();
+            self.write_iter.next().unwrap();
+            return Ok(Some((key, write)));
+        }
+        Ok(None)
+    }
+
+    fn scan_next_batch(&mut self, batch_size: usize) -> Result<Batch> {
+        let mut writes = Vec::with_capacity(batch_size);
+        let mut has_more = true;
+        for _ in 0..batch_size {
+            if let Some((key, write)) = self.next_write()? {
+                let commit_ts = box_try!(Key::decode_ts_from(keys::origin_key(&key)));
+                if commit_ts > self.ts {
+                    writes.push((key, write));
+                }
+            } else {
+                has_more = false;
+                break;
+            }
+        }
+        Ok(Batch { writes, has_more })
+    }
+
+    pub fn process_next_batch(
+        &mut self,
+        batch_size: usize,
+        wb: &mut RocksWriteBatch,
+    ) -> Result<bool> {
+        let Batch { writes, has_more } = self.scan_next_batch(batch_size)?;
+        for (key, write) in writes {
+            let default_key = Key::from_encoded_slice(&key)
+                .truncate_ts()
+                .unwrap()
+                .append_ts(write.start_ts);
+            box_try!(wb.delete_cf(CF_WRITE, &key));
+            box_try!(wb.delete_cf(CF_DEFAULT, default_key.as_encoded()));
+        }
+        wb.write().unwrap();
+        wb.clear();
+        Ok(has_more)
+    }
+
+    pub fn process_next_batch_lock(
+        &mut self,
+        batch_size: usize,
+        wb: &mut RocksWriteBatch,
+    ) -> Result<bool> {
+        let mut has_more = true;
+        for _ in 0..batch_size {
+            if self.lock_iter.valid().unwrap() {
+                let mut state = self
+                    .state
+                    .lock()
+                    .expect("failed to lock ResetToVersionWorker::state");
+                debug_assert!(matches!(
+                    *state,
+                    ResetToVersionState::RemoveLock { scanned: _ }
+                ));
+                *state.scanned() += 1;
+                drop(state);
+
+                box_try!(wb.delete_cf(CF_LOCK, self.lock_iter.key()));
```

Comment:
where is the version check for lock cf?

Reply:
> where is the version check for lock cf?

It's not necessary to do version check for locks, since we need to resolve all locks sooner or later.

/cc https://github.com/tikv/rfcs/pull/81#discussion_r763651376

Reply:
ok.